mcp/tool.go

```diff
@@ -35,40 +35,44 @@ type serverTool struct {
 	inputResolved, outputResolved *jsonschema.Resolved
 }
 
-// newServerTool creates a serverTool from a tool and a handler.
-// If the tool doesn't have an input schema, it is inferred from In.
-// If the tool doesn't have an output schema and Out != any, it is inferred from Out.
-func newServerTool[In, Out any](t *Tool, h ToolHandlerFor[In, Out]) (*serverTool, error) {
-	st := &serverTool{tool: t}
+// A TypedToolHandler handles a call to tools/call with typed arguments and results.
+type TypedToolHandler[In, Out any] func(context.Context, *ServerRequest[*CallToolParams], In) (*CallToolResult, Out, error)
```

Comment:
My 2p: I'm not sure we shouldn't still have CallToolParamsFor[In] and CallToolResultFor[Out], though we wouldn't want the ordinary CallToolParams to be an alias (because @rsc found that confusing).

The important changes seemed to be:
1. That CallToolParams and CallToolResult weren't aliases
2. Making it possible to access the underlying ToolHandler

You can achieve both of those in this CL without adding the additional arguments.
I don't like that there are two ways to access the structured input, and two ways to set the structured output.